lib/config/providers/chat_stream_provider.dart

```diff
@@ -14,70 +16,178 @@ typedef GroupMessageSubscriber =
-            PubkeyFormatter(pubkey: user.publicKey).toHex() ?? '': user,
-        };
-
-        final convertedMessages = await MessageConverter.fromChatMessageList(
-          sortedMessages,
-          currentUserPublicKey: activePubkey,
-          groupId: groupId,
-          usersMap: usersMap,
-        );
-
-        yield convertedMessages;
-      }
+      _rustStreamSubscription = stream.listen(
+        (item) {
+          item.when(
+            initialSnapshot: (messages) {
+              _messageMap = {for (var message in messages) message.id: message};
+            },
+            update: (update) {
+              _messageMap[update.message.id] = update.message;
+            },
+          );
+
+          _emitMergedState(groupId, activePubkey);
+        },
+        onError: (error) {
+          _logger.severe('ChatStreamNotifier: Error in Rust stream', error);
+          if (_messageMap.isEmpty &&
+              _optimisticMessages.isEmpty &&
+              _controller?.isClosed == false) {
+            _controller?.add([]);
+          }
+        },
+      );
     } catch (e) {
       _logger.severe('ChatStreamNotifier: Error building stream for group', e);
-      if (messageMap.isEmpty) {
-        yield [];
+      if (_controller?.isClosed == false) {
+        _controller?.add([]);
       }
     }
   }
+
+  Future<void> addOptimisticMessage(MessageModel message) async {
+    if (_optimisticMessages.any((m) => m.id == message.id)) {
+      return;
+    }
+
+    _optimisticMessages.add(message);
+    final groupId = arg;
+    final activePubkey = ref.read(activePubkeyProvider);
+
+    if (activePubkey != null && activePubkey.isNotEmpty) {
+      await _emitMergedState(groupId, activePubkey);
+    }
+  }
+
+  Future<void> updateOptimisticReaction(MessageModel optimisticReactionMessage) async {
+    final activePubkey = ref.read(activePubkeyProvider);
+
+    if (activePubkey != null && activePubkey.isNotEmpty) {
+      _optimisticMessages.removeWhere((m) => m.id == optimisticReactionMessage.id);
+      _optimisticMessages.add(optimisticReactionMessage);
+
+      final groupId = arg;
+      await _emitMergedState(groupId, activePubkey);
+    }
+  }
+
+  Future<void> removeOptimisticMessage(String messageId) async {
+    _optimisticMessages.removeWhere((m) => m.id == messageId);
+
+    final groupId = arg;
+    final activePubkey = ref.read(activePubkeyProvider);
+
+    if (activePubkey != null && activePubkey.isNotEmpty) {
+      await _emitMergedState(groupId, activePubkey);
+    }
+  }
+
+  Future<void> _emitMergedState(String groupId, String activePubkey) async {
+    if (_isComputingState) {
+      _computeStatePending = true;
+      return;
+    }
+
+    _isComputingState = true;
+    _computeStatePending = false;
+
+    try {
+      await _buildAndEmitState(groupId, activePubkey);
+    } finally {
+      _isComputingState = false;
+      if (_computeStatePending) {
+        await _emitMergedState(groupId, activePubkey);
```

Comment:
We are at a risk of deep recursion here: 
This could lead to potential issues: 
- Under frequent updates, this can recurse repeatedly leading -> stack growth. 

```suggestion
do { ... } while (_computeStatePending);
```